core/internal/leet/config.go

```
@@ -0,0 +1,267 @@
+			_ = os.MkdirAll(dir, 0755)
+		}
+		return m.save()
+	}
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(data, &m.config); err != nil {
+		return err
+	}
+
+	m.normalizeConfig()
+
+	return nil
+}
+
+// normalizeConfig ensures all config values are within valid ranges.
+func (m *ConfigManager) normalizeConfig() {
+	// Clamp grid dimensions
+	m.config.MetricsGrid.Rows = clamp(m.config.MetricsGrid.Rows, MinGridSize, MaxGridSize)
+	m.config.MetricsGrid.Cols = clamp(m.config.MetricsGrid.Cols, MinGridSize, MaxGridSize)
+	m.config.SystemGrid.Rows = clamp(m.config.SystemGrid.Rows, MinGridSize, MaxGridSize)
+	m.config.SystemGrid.Cols = clamp(m.config.SystemGrid.Cols, MinGridSize, MaxGridSize)
+
+	if m.config.HeartbeatInterval <= 0 {
+		m.config.HeartbeatInterval = DefaultHeartbeatInterval
+	}
+}
+
+func clamp(val, min, max int) int {
+	if val < min {
+		return min
+	}
+	if val > max {
+		return max
+	}
+	return val
+}
+
+// save writes the current configuration to disk.
+//
+// Must be called with lock held.
+func (m *ConfigManager) save() error {
+	data, err := json.MarshalIndent(m.config, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	targetPath := m.path
+	tempPath := targetPath + ".tmp"
+
+	// Write atomically via temp file + rename
+	if err := os.WriteFile(tempPath, data, 0644); err != nil {
+		return fmt.Errorf("config: failed to write temp config file: %v", err)
+	}
+	if err := os.Rename(tempPath, targetPath); err != nil {
+		return fmt.Errorf("config: failed to rename tmp config file: %v", err)
+	}
+
+	return nil
+}
+
+// MetricsGrid returns the metrics grid configuration.
+func (m *ConfigManager) MetricsGrid() (rows, cols int) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	return m.config.MetricsGrid.Rows, m.config.MetricsGrid.Cols
+}
+
+// SetMetricsRows sets the metrics grid rows.
+func (m *ConfigManager) SetMetricsRows(rows int) error {
+	if rows < MinGridSize || rows > MaxGridSize {
+		return nil // silently ignore invalid values
```

Why fail silently? Would it be better to notify the user?